packages/overture-schema-core/src/overture/schema/core/scoping/__init__.py

```diff
@@ -0,0 +1,394 @@
+>>> from overture.schema.system.primitive import float32
+...
+>>> @scoped(Scope.VEHICLE)
+... class Fare(BaseModel):
+...     value: float32
+...
+>>> fare_schedule: list[Fare] = [
+...     Fare(
+...         value=10,
+...         when=Fare.When(vehicle=[
+...             VehicleAxleCountSelector(
+...                 dimension=VehicleDimension.AXLE_COUNT,
+...                 comparison=VehicleRelation.LESS_THAN,
+...                 value=3
+...             ),
+...             VehicleLengthSelector(
+...                dimension=VehicleDimension.LENGTH,
+...                comparison=VehicleRelation.LESS_THAN_EQUAL,
+...                value=18,
+...                unit=LengthUnit.FT,
+...             )
+...         ]),
+...     ),
+...     Fare(value=30),
+... ]
+
+
+
+Mixing scopes
+==============
+The `@scoped` decorator can mix any desired combination of scopes onto your Pydantic model.
+
+For example, suppose you are modeling a value type that can apply at certain times, along a certain
+sub-segment of a linear path, depending on whether one is traveling forward or backward along the
+path, or any combination of these three. This can be achieved easily with:
+
+>>> @scoped(Scope.TEMPORAL, Scope.GEOMETRIC_RANGE, Scope.HEADING)
+... class MyModel(BaseModel):
+...     pass
+
+
+Optional and required scopes
+============================
+When using the `@scoped` decorator, scopes are optional by default but some or all scopes may be
+made required.
+
+The following example makes all the scopes required:
+
+>>> from enum import Enum
+...
+>>> class SignalType(str, Enum):
+...    STOP_SIGN = 'stop_sign'
+...
+>>> @scoped(required=(Scope.GEOMETRIC_POSITION, Scope.HEADING))
+... class TrafficSignal(BaseModel):
+...    signal_type: SignalType
+
+The following example mixes an optional scope (temporal) with two required scopes (geometric
+position and heading).
+
+>>> @scoped(Scope.TEMPORAL, required=(Scope.GEOMETRIC_POSITION, Scope.HEADING))
+... class TrafficSignal(BaseModel):
+...    signal_type: SignalType
+
+
+The `when` clause
+=================
+For historical reasons, some scope fields are added directly to the decorated model, while others
+are added as children of a synthetic `when` field.
+
+| Scope                      | Field             |
+|----------------------------|-------------------|
+| `Scope.GEOMETRIC_POSITION` | `at`              |
+| `Scope.GEOMETRIC_RANGE`    | `between`         |
+| `Scope.HEADING`            | `when.heading`    |
+| `Scope.TEMPORAL`           | `when.during`     |
+| `Scope.TRAVEL_MODE`        | `when.mode`       |
+| `Scope.PURPOSE_OF_USE`     | `when.using`      |
+| `Scope.RECOGNIZED_STATUS`  | `when.recognized` |
+| `Scope.SIDE`               | `side`            |
+| `Scope.VEHICLE`            | `when.vehicle`    |
+
+
+If a `when` field is added to the model, the model is also decorated with a nested `When` class to
+simplify instantiating values for the `when` field, for example:
```

Comment:
These answered questions I had above. I'd open with this to provide more of an introduction before digging into specific concepts.

What _is_ the nested `When` class? Reading this, I don't know what the right mental model is, which is relevant because I might try to duck type my way through with a `dict` (see above).